pool/src/transaction_pool.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A minimal (EXPERIMENTAL) transaction pool implementation
+
+use std::sync::Arc;
+use time;
+
+use core::core::transaction;
+use core::core::{Block, CompactBlock, Transaction};
+use pool::Pool;
+use types::*;
+
+/// A minimal (EXPERIMENTAL) transaction pool implementation
+pub struct TransactionPool<T> {
+	/// Pool Config
+	pub config: PoolConfig,
+
+	/// Our transaction pool.
+	pub txpool: Pool<T>,
+	/// Our Dandelion "stempool".
+	pub stempool: Pool<T>,
+
+	/// The blockchain
+	pub blockchain: Arc<T>,
+	/// The pool adapter
+	pub adapter: Arc<PoolAdapter>,
+}
+
+impl<T> TransactionPool<T>
+where
+	T: BlockChain,
+{
+	/// Create a new transaction pool
+	pub fn new(config: PoolConfig, chain: Arc<T>, adapter: Arc<PoolAdapter>) -> TransactionPool<T> {
+		TransactionPool {
+			config: config,
+			txpool: Pool::new(chain.clone(), format!("txpool")),
+			stempool: Pool::new(chain.clone(), format!("stempool")),
+			blockchain: chain,
+			adapter: adapter,
+		}
+	}
+
+	fn add_to_stempool(&mut self, entry: PoolEntry) -> Result<(), PoolError> {
+		// Add tx to stempool (passing in all txs from txpool to validate against).
+		self.stempool
+			.add_to_pool(entry.clone(), self.txpool.all_transactions())?;
+
+		// Note: we do not notify the adapter here,
+		// we let the dandelion monitor handle this.
+		Ok(())
+	}
+
+	fn add_to_txpool(&mut self, mut entry: PoolEntry) -> Result<(), PoolError> {
+		// First deaggregate the tx based on current txpool txs.
+		if entry.tx.kernels.len() > 1 {
+			let txs = self.txpool
+				.find_matching_transactions(entry.tx.kernels.clone());
+			if !txs.is_empty() {
+				entry.tx = transaction::deaggregate(entry.tx, txs)?;
+				entry.src.debug_name = "deagg".to_string();
+			}
+		}
+		self.txpool.add_to_pool(entry.clone(), vec![])?;
+
+		// We now need to reconcile the stempool based on the new state of the txpool.
+		// Some stempool txs may no longer be valid and we need to evict them.
+		let txpool_tx = self.txpool.aggregate_transaction()?;
+		self.stempool.reconcile(txpool_tx)?;
+
+		self.adapter.tx_accepted(&entry.tx);
+		Ok(())
+	}
+
+	/// Add the given tx to the pool, directing it to either the stempool or
+	/// txpool based on stem flag provided.
+	pub fn add_to_pool(
+		&mut self,
+		src: TxSource,
+		tx: Transaction,
+		stem: bool,
```

Comment:
Unrelated, I saw the same flag in the previous version, but this boolean flag doesn't look good, perhaps we should refactor later on to make sure that method has one responsibility

Reply:
Yeah we can definitely revisit this. I went back and forth on this with the flag vs two separate functions. Landed on this for now to keep it closer to existing impl.